### PR TITLE
[GBM] implement an atomic queue

### PR DESCRIPTION
**Pull request by lrusak**
This adds an atomic queue to the atomic DRM windowing. This allows us to avoid having the gui possibly enter an unusable state in the case of a bad request.

I've implemented this as a stack using `std::deque` (I didn't use `std::stack` because I need to access the other elements).

So if there is an unsuccessful request there should also be a "successful" request in the stack. The successful request only gets removed when another successful request is added. This allows us to fallback to the successful request in case there is an unsuccessful request. I hope this makes sense.

This allows us to fallback gracefully in case the atomic (test) commit fails. The only thing that will be updated is the fb id (this allows us to avoid tearing and keep up with the swapchain).

I've also implemented a custom class that wraps the `drmModeAtomicReq` object. This is so we can track the atomic request with all the objects, properties, and values being submitted. We want to do this so we can either log the info (if `LOGWINDOWING` component logging is enabled) or in case the request fails we can log the difference between the two requests to see what is different. The only real way to check what cause the failure is to look at the kernel logs with atomic drm logging enabled.

The logging looks like this:

----

diff
```
2020-10-19 09:31:50.340 T:2320878   DEBUG <general>: CDRMAtomicRequest::LogAtomicDiff - DRM Atomic Request Diff:
2020-10-19 09:31:50.340 T:2320878   ERROR <general>:
                                                   Object: plane        ID: 45
                                                     Property: CRTC_W   ID: 15  Value: 18446744073709551615
                                                     Property: CRTC_H   ID: 16  Value: 18446744073709551615
                                                     Property: FB_ID    ID: 17  Value: 100
```
For this failed request I forced a value of `-1` to the `CRTC_W` and `CRTC_H` This cause a failure because it expects a uint64_t value. The FB_ID is shown here also because that will be different than the last request as the swapchain rotates.

I used this commit to test this: https://github.com/lrusak/xbmc/commit/b52a8413960a748dc62755a01ad17dff5ca708a3

----

full request
```
2020-10-19 09:11:03.447 T:2314219   DEBUG <general>: CDRMAtomicRequest::LogAtomicRequest - DRM Atomic Request:
2020-10-19 09:11:03.447 T:2314219   DEBUG <general>:
                                                   Object: plane        ID: 45
                                                     Property: SRC_X    ID: 9   Value: 0
                                                     Property: SRC_Y    ID: 10  Value: 0
                                                     Property: SRC_W    ID: 11  Value: 125829120
                                                     Property: SRC_H    ID: 12  Value: 70778880
                                                     Property: CRTC_X   ID: 13  Value: 0
                                                     Property: CRTC_Y   ID: 14  Value: 0
                                                     Property: CRTC_W   ID: 15  Value: 1920
                                                     Property: CRTC_H   ID: 16  Value: 1080
                                                     Property: FB_ID    ID: 17  Value: 97
                                                     Property: CRTC_ID  ID: 20  Value: 47
```

I don't love the indirection of the `UpdateRequestQueue` method but I feel it's still better to have as a helper function to make it more clear what is happening if the request is successful or unsuccessful.

### CODE REVIEW

File: xbmc/windowing/gbm/drm/DRMAtomic.h [hunk 1]
```diff
@@ -10,6 +10,10 @@
 
 #include "DRMUtils.h"
 
+#include <deque>
+#include <map>
+#include <memory>
```
Comment (garbear) [inline xbmc/windowing/gbm/drm/DRMAtomic.h:15]: `uint32_t` and `uint64_t` are added, so `<stdint.h>` should be included here

File: xbmc/windowing/gbm/drm/DRMAtomic.cpp [hunk 2]
```diff
@@ -137,30 +137,36 @@ void CDRMAtomic::DrmAtomicCommit(int fb_id, int flags, bool rendered, bool video
...
   if (flags & DRM_MODE_ATOMIC_ALLOW_MODESET)
   {
     if (drmModeDestroyPropertyBlob(m_fd, blob_id) != 0)
       CLog::Log(LOGERROR, "CDRMAtomic::{} - failed to destroy property blob: {}", __FUNCTION__,
                 strerror(errno));
   }
 
-  drmModeAtomicFree(m_req);
-  m_req = drmModeAtomicAlloc();
+  UpdateRequestQueue(true);
```
Comment (a1rwulf) [inline xbmc/windowing/gbm/drm/DRMAtomic.cpp:169]: I struggled a bit to find this initialization.
Imho it's not quite obvious that updating the queue initializes `m_req`.
How about explicitly initializing it?

File: xbmc/windowing/gbm/drm/DRMAtomic.h (hunk 1 above)
Comment (lrusak) [reply xbmc/windowing/gbm/drm/DRMAtomic.h:16]: Thanks!

File: xbmc/windowing/gbm/drm/DRMAtomic.cpp [hunk 3]
```diff
@@ -256,23 +259,103 @@ bool CDRMAtomic::SetActive(bool active)
...
+    m_req->LogAtomicDiff(m_atomicRequestQueue.front().get());
+
+  if (m_atomicRequestQueue.size() > 1)
+    m_atomicRequestQueue.pop_back();
+
+  if (success)
+    m_atomicRequestQueue.emplace_back(std::make_unique<CDRMAtomicRequest>());
+
+  m_req = m_atomicRequestQueue.back().get();
+}
```
[the remark this answers is not in the document]
Comment (lrusak) [reply xbmc/windowing/gbm/drm/DRMAtomic.cpp]: I agree 😄 That's why I said in the OP that I didn't love this method.

I've removed this method in the latest changes and tried to make it more clear about what is happening and why.

I've also changed the LogAtomicDiff method static and accept two CDRMAtomicRequest objects to clarify that it is doing a diff of the two requests.

File: xbmc/windowing/gbm/drm/DRMAtomic.cpp (hunk 2 above)
Comment (lrusak) [reply xbmc/windowing/gbm/drm/DRMAtomic.cpp]: Yep, updated thanks!

File: xbmc/windowing/gbm/drm/DRMAtomic.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (a1rwulf) [reply xbmc/windowing/gbm/drm/DRMAtomic.cpp]: LOL, I totally overread that one.
Anyway, I think the code is more clear now.
1 question remaining:
Do you even need the m_req member at all, it's just a pointer the back of the queue.
As this is only local to one method, you could store the pointer in a local variable and don't use m_req at all.
But I'll leave that up to you.

**Review by a1rwulf — APPROVED** [review]

File: xbmc/windowing/gbm/drm/DRMAtomic.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (lrusak) [reply xbmc/windowing/gbm/drm/DRMAtomic.cpp]: m_req is also used in the `AddProperty` method. In the future I am probably going to remove this though and add a getter for the atomic request `m_req` object

[the remark this answers is not in the document]
Comment (a1rwulf) [reply xbmc/windowing/gbm/drm/DRMAtomic.cpp]: I see, all fine then.

**Comment by lrusak** [comment]
jenkins build this please

**Comment by lrusak** [comment]
jenkins build this please

**Pull request merged**